src/compiler/checker.ts

```diff
@@ -38232,7 +38233,7 @@ namespace ts {
                     ? condExpr.right
                     : condExpr;
                 if (isModuleExportsAccessExpression(location)) return;
-                const type = checkTruthinessExpression(location);
+                const type = location === condExpr ? condType : checkTruthinessExpression(location);

```

Comment:
If anyone has ideas on how to avoid this `checkTruthinessExpression` altogether, please let me know.

Reply:
Yeah, that seems tricky because it looks like `location` is either `condExpr`, and we already have the type, or it is `condExpr.right`, and then... we haven't visited yet in the trampoline on `onOperator`, right?

Reply:
Indeed, this is my understanding as well. And I don't think that a dynamic programming like approach is worth it here, given the overhead of caching the type of subtrees. Given that the performance issue that manifested itself is mitigated by the current changes I'd consider it fine that we leave this like this, at least until another performance bottleneck is observed.